File: system-test/dotnet/Driver.Adapter/Shared/Client/Http/JsonHttpClient.cs
using System.Net;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Common;

namespace Driver.Adapter.Shared.Client.Http;

public class JsonHttpClient<E> : IDisposable
{
    private const string ApplicationJson = "application/json";

    private readonly JsonSerializerOptions _jsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        NumberHandling = JsonNumberHandling.AllowReadingFromString,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase, allowIntegerValues: false) }
    };

    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private bool _disposed;

    public JsonHttpClient(string baseUrl)
    {
        _httpClient = new HttpClient { BaseAddress = new Uri(baseUrl) };
        _baseUrl = baseUrl;
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                _httpClient?.Dispose();
            }
            _disposed = true;
        }
    }

    public async Task<Result<T, E>> GetAsync<T>(string path)
        => await GetResultOrFailureAsync<T>(await DoGetAsync(path));

    public async Task<Result<VoidValue, E>> GetAsync(string path)
        => await GetResultOrFailureAsync<VoidValue>(await DoGetAsync(path));

    public async Task<Result<T, E>> PostAsync<T>(string path, object request)
        => await GetResultOrFailureAsync<T>(await DoPostAsync(path, request));

    public async Task<Result<VoidValue, E>> PostAsync(string path, object request)
        => await GetResultOrFailureAsync<VoidValue>(await DoPostAsync(path, request));

    public async Task<Result<VoidValue, E>> PostAsync(string path)
        => await GetResultOrFailureAsync<VoidValue>(await DoPostAsync(path));

    public async Task<Result<T, E>> PutAsync<T>(string path, object request)
        => await GetResultOrFailureAsync<T>(await DoPutAsync(path, request));

    public async Task<Result<VoidValue, E>> PutAsync(string path, object request)
        => await GetResultOrFailureAsync<VoidValue>(await DoPutAsync(path, request));

    public async Task<Result<T, E>> DeleteAsync<T>(string path)
        => await GetResultOrFailureAsync<T>(await DoDeleteAsync(path));

    public async Task<Result<VoidValue, E>> DeleteAsync(string path)
        => await GetResultOrFailureAsync<VoidValue>(await DoDeleteAsync(path));

    private async Task<HttpResponseMessage> DoGetAsync(string path)
    {
        var uri = GetUri(path);
        var httpRequest = new HttpRequestMessage(HttpMethod.Get, uri);
        return await SendRequest(httpRequest);
    }

    #region Helpers

    private Uri GetUri(string path)
    {
        return new Uri(_baseUrl + path);
    }

    private async Task<HttpResponseMessage> DoPostAsync(string path, object request)
    {
        var uri = GetUri(path);
        var jsonBody = SerializeRequest(request);
        var httpRequest = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = new StringContent(jsonBody, Encoding.UTF8, ApplicationJson)
        };
        return await SendRequest(httpRequest);
    }

    private async Task<HttpResponseMessage> DoPostAsync(string path)
    {
        var uri = GetUri(path);
        var httpRequest = new HttpRequestMessage(HttpMethod.Post, uri)
        {
            Content = new StringContent(string.Empty, Encoding.UTF8, ApplicationJson)
        };
        return await SendRequest(httpRequest);
    }

    private async Task<HttpResponseMessage> DoPutAsync(string path, object request)
    {
        var uri = GetUri(path);
        var jsonBody = SerializeRequest(request);
        var httpRequest = new HttpRequestMessage(HttpMethod.Put, uri)
        {
            Content = new StringContent(jsonBody, Encoding.UTF8, ApplicationJson)
        };
        return await SendRequest(httpRequest);
    }

    private async Task<HttpResponseMessage> DoDeleteAsync(string path)
    {
        var uri = GetUri(path);
        var httpRequest = new HttpRequestMessage(HttpMethod.Delete, uri);
        return await SendRequest(httpRequest);
    }

    private Task<HttpResponseMessage> SendRequest(HttpRequestMessage httpRequest)
        => _httpClient.SendAsync(httpRequest);

    private string SerializeRequest(object request)
    {
        return JsonSerializer.Serialize(request, _jsonOptions);
    }

    private static async Task<T> ReadResponseAsync<T>(HttpResponseMessage httpResponse, JsonSerializerOptions jsonOptions)
    {
        var responseBody = await httpResponse.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<T>(responseBody, jsonOptions)!;
    }

    private async Task<Result<T, E>> GetResultOrFailureAsync<T>(HttpResponseMessage httpResponse)
    {
        if (!httpResponse.IsSuccessStatusCode)
        {
            var error = await ReadResponseAsync<E>(httpResponse, _jsonOptions);
            return Result<T, E>.Failure(error);
        }

        if (typeof(T) == typeof(VoidValue) || httpResponse.StatusCode == HttpStatusCode.NoContent)
        {
            return Result<T, E>.Success(default!);
        }

        var response = await ReadResponseAsync<T>(httpResponse, _jsonOptions);
        return Result<T, E>.Success(response);
    }

    #endregion
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/shared/client/http/JsonHttpClient.java
package com.optivem.shop.dsl.driver.adapter.shared.client.http;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.common.Result;
import org.springframework.http.HttpStatus;

import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class JsonHttpClient<E> implements AutoCloseable {
    private static final String CONTENT_TYPE = "Content-Type";
    private static final String APPLICATION_JSON = "application/json";
    private static final ExecutorService VIRTUAL_THREAD_EXECUTOR = Executors.newVirtualThreadPerTaskExecutor();
    private final ObjectMapper objectMapper;

    private final HttpClient httpClient;
    private final String baseUrl;
    private final Class<E> errorType;

    public JsonHttpClient(HttpClient httpClient, String baseUrl, Class<E> errorType, ObjectMapper objectMapper) {
        this.httpClient = httpClient;
        this.baseUrl = baseUrl;
        this.errorType = errorType;
        this.objectMapper = objectMapper;
    }

    public JsonHttpClient(String baseUrl, Class<E> errorType) {
        this(HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1).build(), baseUrl, errorType, createObjectMapper());
    }

    private static ObjectMapper createObjectMapper() {
        var mapper = new ObjectMapper();
        mapper.registerModule(new JavaTimeModule());
        mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        // Enable case-insensitive property binding
        mapper.setConfig(mapper.getDeserializationConfig()
                .with(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES));
        return mapper;
    }

    @Override
    public void close() {
        Closer.close(httpClient);
    }

    // GET Methods

    public <T> Result<T, E> get(String path, Class<T> responseType) {
        var httpResponse = doGet(path);
        return getResultOrFailure(httpResponse, responseType);
    }

    public Result<Void, E> get(String path) {
        var httpResponse = doGet(path);
        return getResultOrFailure(httpResponse, Void.class);
    }
    
    // POST Methods
    
    public <T> Result<T, E> post(String path, Object request, Class<T> responseType) {
        var httpResponse = doPost(path, request);
        return getResultOrFailure(httpResponse, responseType);
    }
    
    public Result<Void, E> post(String path, Object request) {
        var httpResponse = doPost(path, request);
        return getResultOrFailure(httpResponse, Void.class);
    }

    public <T> Result<T, E> post(String path, Class<T> responseType) {
        var httpResponse = doPost(path);
        return getResultOrFailure(httpResponse, responseType);
    }

    public Result<Void, E> post(String path) {
        var httpResponse = doPost(path);
        return getResultOrFailure(httpResponse, Void.class);
    }

    // PUT Methods

    public <T> Result<T, E> put(String path, Object request, Class<T> responseType) {
        var httpResponse = doPut(path, request);
        return getResultOrFailure(httpResponse, responseType);
    }

    public Result<Void, E> put(String path, Object request) {
        var httpResponse = doPut(path, request);
        return getResultOrFailure(httpResponse, Void.class);
    }

    // DELETE Methods

    public <T> Result<T, E> delete(String path, Class<T> responseType) {
        var httpResponse = doDelete(path);
        return getResultOrFailure(httpResponse, responseType);
    }

    public Result<Void, E> delete(String path) {
        var httpResponse = doDelete(path);
        return getResultOrFailure(httpResponse, Void.class);
    }

    // Private helper methods

    private URI getUri(String path) {
        try {
            return new URI(baseUrl + path);
        } catch (URISyntaxException ex) {
            throw new IllegalArgumentException("Failed to create URI for path: " + path, ex);
        }
    }

    private HttpResponse<String> doGet(String path) {
        var uri = getUri(path);
        var httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .GET()
                .build();

        return sendRequest(httpRequest);
    }

    private HttpResponse<String> doPost(String path, Object request) {
        var uri = getUri(path);
        var jsonBody = serializeRequest(request);

        var httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .header(CONTENT_TYPE, APPLICATION_JSON)
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();

        return sendRequest(httpRequest);
    }

    private HttpResponse<String> doPost(String path) {
        var uri = getUri(path);

        var httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .header(CONTENT_TYPE, APPLICATION_JSON)
                .POST(HttpRequest.BodyPublishers.noBody())
                .build();

        return sendRequest(httpRequest);
    }

    private HttpResponse<String> doPut(String path, Object request) {
        var uri = getUri(path);
        var jsonBody = serializeRequest(request);

        var httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .header(CONTENT_TYPE, APPLICATION_JSON)
                .PUT(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();

        return sendRequest(httpRequest);
    }

    private HttpResponse<String> doDelete(String path) {
        var uri = getUri(path);

        var httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .DELETE()
                .build();

        return sendRequest(httpRequest);
    }

    private HttpResponse<String> sendRequest(HttpRequest httpRequest) {
        try {
            // Execute HTTP request on virtual thread for non-blocking I/O
            return VIRTUAL_THREAD_EXECUTOR.submit(() ->
                    httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())
            ).get();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Failed to send HTTP request", e);
        } catch (ExecutionException e) {
            var cause = e.getCause();
            throw new IllegalStateException("Failed to send HTTP request", cause != null ? cause : e);
        }
    }

    private String serializeRequest(Object request) {
        try {
            return objectMapper.writeValueAsString(request);
        } catch (com.fasterxml.jackson.core.JsonProcessingException ex) {
            throw new IllegalStateException("Failed to serialize request object", ex);
        }
    }

    private <T> T readResponse(HttpResponse<String> httpResponse, Class<T> responseType) {
        try {
            var responseBody = httpResponse.body();
            return objectMapper.readValue(responseBody, responseType);
        } catch (com.fasterxml.jackson.core.JsonParseException ex) {
            var responseBody = httpResponse.body();
            var fallback = tryCreateErrorFromBody(responseType, responseBody, ex);
            if (fallback != null) {
                return fallback;
            }
            throw new IllegalStateException("Failed to deserialize response to " + responseType.getSimpleName() + ". Response body: " + responseBody, ex);
        } catch (com.fasterxml.jackson.core.JsonProcessingException ex) {
            throw new IllegalStateException("Failed to deserialize response to " + responseType.getSimpleName(), ex);
        }
    }

    @SuppressWarnings("java:S3011") // Reflection needed for error response fallback when JSON parsing fails
    private <T> T tryCreateErrorFromBody(Class<T> responseType, String responseBody, com.fasterxml.jackson.core.JsonParseException ex) {
        try {
            var constructor = responseType.getDeclaredConstructor();
            constructor.setAccessible(true);
            var instance = constructor.newInstance();
            var messageField = responseType.getDeclaredField("message");
            messageField.setAccessible(true);
            messageField.set(instance, responseBody);
            return instance;
        } catch (ReflectiveOperationException e) {
            return null;
        }
    }

    private boolean isSuccessStatusCode(HttpResponse<String> httpResponse) {
        var statusCode = HttpStatus.valueOf(httpResponse.statusCode());
        return statusCode.is2xxSuccessful();
    }

    private <T> Result<T, E> getResultOrFailure(HttpResponse<String> httpResponse, Class<T> responseType) {
        if (!isSuccessStatusCode(httpResponse)) {
            var error = readResponse(httpResponse, errorType);
            return Result.failure(error);
        }

        if (responseType == Void.class || httpResponse.statusCode() == HttpStatus.NO_CONTENT.value()) {
            return Result.success(null);
        }

        var response = readResponse(httpResponse, responseType);
        return Result.success(response);
    }
}
